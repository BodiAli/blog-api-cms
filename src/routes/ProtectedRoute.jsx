import { useState, useEffect } from "react";
import { Navigate } from "react-router";
import { toast } from "react-toastify";
import { UserContext } from "../utils/UserContext";
import Loader from "../components/Loader/Loader";

export default function ProtectedRoute({ children }) {
  const [user, setUser] = useState(null);
  const [redirect, setRedirect] = useState(false);
  const [loading, setLoading] = useState(true);

  const token = localStorage.getItem("token");

  useEffect(() => {
    if (!token) {
      toast.error("You need to sign in or sign up before continuing.");
      setLoading(false);
      setRedirect(true);
      return;
    }

    async function fetchUser() {
      try {
        const res = await fetch(`${import.meta.env.VITE_SERVER_URL}/auth/validate`, {
          headers: {
            Authorization: token,
          },
        });

        if (!res.ok) {
          if (res.status === 401) {
            toast.error("You need to sign in or sign up before continuing.");
            setRedirect(true);
            return;
          }
          throw new Error("Failed to fetch user");
        }

        const fetchedUser = await res.json();

        setUser(fetchedUser);
      } catch (error) {
        toast.error(error.message || "Error fetching user");
        setUser(null);
        setRedirect(true);
      } finally {
        setLoading(false);
      }
    }

    fetchUser();
  }, [token]);

  if (loading) return <Loader />;
  console.log("USER", user);

  if (redirect) {
    return <Navigate to="log-in" replace={true} />;
  }

  return <UserContext.Provider value={user}>{children}</UserContext.Provider>;
}
