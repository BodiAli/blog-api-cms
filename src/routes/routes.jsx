import App from "../App";
import ProtectedRoute from "./ProtectedRoute.jsx";
import Signup from "../pages/Signup.jsx";
import Login from "../pages/Login.jsx";
import NotFoundPage from "../pages/NotFoundPage.jsx";

const routes = [
  {
    path: "/",
    element: (
      <ProtectedRoute>
        <App />
      </ProtectedRoute>
    ),
    errorElement: <NotFoundPage />,
  },

  {
    path: "log-in",
    Component: Login,
  },

  {
    path: "sign-up",
    Component: Signup,
  },
];

export default routes;
