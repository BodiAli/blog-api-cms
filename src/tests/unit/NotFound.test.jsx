import { render } from "@testing-library/react";
import { createMemoryRouter, RouterProvider } from "react-router";
import routes from "../../routes/routes";

describe("NotFound component", () => {
  test("Should render NotFound component with link to head back to '/' path", async () => {
    const router = createMemoryRouter(routes, { initialEntries: ["/undefined-path"] });

    const { container } = render(<RouterProvider router={router} />);

    expect(container).toMatchSnapshot();
  });
});
