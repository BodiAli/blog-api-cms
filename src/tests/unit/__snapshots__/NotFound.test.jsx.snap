// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NotFound component > Should render NotFound component with link to head back to '/' path 1`] = `
<div>
  <main
    class="_main_069038"
  >
    <h1>
      404 Not found
    </h1>
    <p>
      Looks like you're lost! 
      <a
        data-discover="true"
        href="/"
      >
        Head back.
      </a>
    </p>
  </main>
</div>
`;
